lib/spack/spack/environment/environment.py

```diff
@@ -1442,6 +1443,59 @@ def is_develop(self, spec):
         """Returns true when the spec is built from local sources"""
         return spec.name in self.dev_specs
 
+    def apply_develop(self, spec: spack.spec.Spec, path: Optional[str] = None):
+        """Mutate concrete specs to include dev_path provenance pointing to path.
+
+        This does not do any other aspect of concretization. It will fail if any existing concrete
+        spec for the same package does not satisfy the given develop spec."""
+        # Find all specs that this develop request applies to
+        modify_specs = []
+        for dep in traverse.traverse_nodes(list(self.specs_by_hash.values())):
+            if dep.name == spec.name:
+                if not dep.satisfies(spec):
+                    msg = f"Develop spec '{spec}' conflicts with concrete specs in environment."
+                    msg += " Try again with 'spack develop --no-apply-changes'"
+                    msg += " and run 'spack concretize --force' to apply changes."
+                    raise SpackEnvironmentDevelopError(msg)
```

Comment:
Style thing, but I usually do `raise Error(<literal string>)` because of python's scoping rules where `msg` is available in the entirety of the function (and can sometimes lead to bugs during refactor)

Reply:
I think I prefer your way, although it's not the more common one across our codebase. I've updated this